store/index/index.go

```diff
@@ -90,119 +92,226 @@ const BucketPoolSize = 1024
+
+	go idx.garbageCollector()
+
+	return idx, nil
+}
+
+func indexFileName(basePath string, fileNum uint32) string {
+	return fmt.Sprintf("%s.%d", basePath, fileNum)
+}
+
+func scanIndex(basePath string, fileNum uint32, indexSizeBits uint8, buckets Buckets, sizeBuckets SizeBuckets, fileBuckets FileBuckets) (uint32, error) {
+	if len(buckets) != 1<<indexSizeBits {
+		panic("wrong size of buckets")
+	}
+	if len(sizeBuckets) != 1<<indexSizeBits {
+		panic("wrong size of sizeBuckets")
+	}
+	if len(fileBuckets) != 1<<indexSizeBits {
+		panic("wrong size of fileBuckets")
+	}
+
+	var lastFileNum uint32
+	for {
+		err := scanIndexFile(basePath, fileNum, indexSizeBits, buckets, sizeBuckets, fileBuckets)
+		if err != nil {
+			if os.IsNotExist(err) {
+				break
+			}
+			return 0, err
+		}
+		lastFileNum = fileNum
+		fileNum++
+	}
+	return lastFileNum, nil
+}
+
+func (i *Index) StorageSize() (int64, error) {
+	header, err := readHeader(i.headerPath)
 	if err != nil {
-		return nil, nil, err
+		if os.IsNotExist(err) {
+			return 0, nil
+		}
+		return 0, err
 	}
-	sizeBuckets, err := NewSizeBuckets(indexSizeBits)
+	var size int64
+	fi, err := os.Stat(i.headerPath)
 	if err != nil {
-		return nil, nil, err
+		return 0, err
 	}
-	buffered := bufio.NewReader(file)
-	iter := NewIndexIter(buffered, types.Position(bytesRead))
+	size += fi.Size()
+
+	fi, err = os.Stat(i.basePath + ".free")
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return 0, err
+		}
+	} else {
+		size += fi.Size()
+	}
+
+	fileNum := header.FirstFile
 	for {
-		data, pos, done, err := iter.Next()
-		if done {
-			break
+		fi, err = os.Stat(indexFileName(i.basePath, fileNum))
+		if err != nil {
+			if os.IsNotExist(err) {
+				break
+			}
+			return 0, err
 		}
-		if err == io.EOF {
-			// The file is corrupt. Though it's not a problem, just take the data we
-			// are able to use and move on.
-			if _, err := file.Seek(0, 2); err != nil {
-				return nil, nil, err
+		size += fi.Size()
+		fileNum++
+	}
+	return size, nil
+}
+
+func scanIndexFile(basePath string, fileNum uint32, indexSizeBits uint8, buckets Buckets, sizeBuckets SizeBuckets, fileBuckets FileBuckets) error {
+	indexPath := indexFileName(basePath, fileNum)
+
+	// This is a single sequential read across the index file.
+	file, err := openFileForScan(indexPath)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	buffered := bufio.NewReaderSize(file, 32*1024)
+	sizeBuffer := make([]byte, SizePrefixSize)
+	scratch := make([]byte, 256)
+	var iterPos int64
+	for {
+		_, err = io.ReadFull(buffered, sizeBuffer)
+		if err != nil {
+			if err == io.EOF {
+				break
 			}
-			break
+			return err
+		}
+		size := binary.LittleEndian.Uint32(sizeBuffer)
+
+		pos := iterPos + SizePrefixSize
+		iterPos = pos + int64(size)
+		if int(size) > len(scratch) {
+			scratch = make([]byte, size)
 		}
+		data := scratch[:size]
+		_, err = io.ReadFull(buffered, data)
 		if err != nil {
-			return nil, nil, err
+			if err == io.EOF {
```

Comment:
Would be great to have a test case for this.

Reply:
OK, but it does not work at all without that.  That is how it knows it has no more files to read.

Reply:
I added a log message here.  If this happens it means some index records could not be read from that file.  At some point the file should get replaced and GC'd.  I am not sure what to test for. It just means that the corrupt index record does not appear in the file